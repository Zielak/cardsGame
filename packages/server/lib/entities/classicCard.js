"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __read = (this && this.__read) || function (o, n) {
    var m = typeof Symbol === "function" && o[Symbol.iterator];
    if (!m) return o;
    var i = m.call(o), r, ar = [], e;
    try {
        while ((n === void 0 || n-- > 0) && !(r = i.next()).done) ar.push(r.value);
    }
    catch (error) { e = { error: error }; }
    finally {
        try {
            if (r && !r.done && (m = i["return"])) m.call(i);
        }
        finally { if (e) throw e.error; }
    }
    return ar;
};
var __spread = (this && this.__spread) || function () {
    for (var ar = [], i = 0; i < arguments.length; i++) ar = ar.concat(__read(arguments[i]));
    return ar;
};
Object.defineProperty(exports, "__esModule", { value: true });
var card_1 = require("./card");
var logs_1 = require("../logs");
var decorators_1 = require("../decorators");
var ClassicCard = /** @class */ (function (_super) {
    __extends(ClassicCard, _super);
    function ClassicCard(options) {
        var _this = _super.call(this, options) || this;
        _this.type = "classicCard";
        _this._visibilityData.add(["name", "suit", "rank"], 
        /* toEveryone */ function () { return _this.faceUp; }, 
        /* toOwner */ function () {
            // Only if it's in his hand
            var parentContainer = _this.parentEntity;
            return parentContainer.type === "hand";
        });
        _this.suit = options.suit;
        _this.rank = options.rank;
        _this.name = _this.rank + _this.suit;
        return _this;
    }
    __decorate([
        decorators_1.condvis
    ], ClassicCard.prototype, "name", void 0);
    __decorate([
        decorators_1.condvis
    ], ClassicCard.prototype, "suit", void 0);
    __decorate([
        decorators_1.condvis
    ], ClassicCard.prototype, "rank", void 0);
    return ClassicCard;
}(card_1.Card));
exports.ClassicCard = ClassicCard;
exports.standardDeck = function (ranks, suits) {
    if (ranks === void 0) { ranks = [
        "2",
        "3",
        "4",
        "5",
        "6",
        "7",
        "8",
        "9",
        "10",
        "J",
        "Q",
        "K",
        "A"
    ]; }
    if (suits === void 0) { suits = ["H", "S", "C", "D"]; }
    var cards = suits.reduce(function (prevS, suit) { return __spread(prevS, ranks.reduce(function (prevR, rank) { return __spread(prevR, [{ suit: suit, rank: rank }]); }, [])); }, []);
    logs_1.logs.verbose("created a deck of " + cards.length + " cards");
    return cards;
};
//# sourceMappingURL=classicCard.js.map