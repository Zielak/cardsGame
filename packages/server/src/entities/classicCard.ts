import { IBaseCardOptions } from "./baseCard"
import { logs } from "../logs"
import { Entity, IEntityImplementation } from "./entity"

export const ClassicCard: IEntityImplementation = (
  entity: Entity,
  options: IClassicCardOptions
) => {
  entity.type = "classicCard"
  entity.data.suit = options.suit
  entity.data.rank = options.rank
  entity.name = options.suit + options.rank

  return {
    suit: {
      get: () => entity.data.suit
    },
    rank: {
      get: () => entity.data.rank
    }
  }
}

export interface IClassicCardOptions extends IBaseCardOptions {
  suit: string
  rank: string
}

export const standardDeck = (
  // prettier-ignore
  ranks: string[] = [
    "2", "3", "4", "5", "6", "7", "8", "9", "10", "J", "Q", "K", "A"
  ],
  suits: string[] = ["H", "S", "C", "D"]
): IClassicCardOptions[] => {
  const cards: IClassicCardOptions[] = suits.reduce(
    (prevS, suit) => [
      ...prevS,
      ...ranks.reduce((prevR, rank) => [...prevR, { suit, rank }], [])
    ],
    []
  )

  logs.verbose(`created a deck of ${cards.length} cards`)

  return cards
}
