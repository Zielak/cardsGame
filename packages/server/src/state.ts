import { nosync } from "colyseus"
import { EventEmitter } from "eventemitter3"
import { logs } from "./logs"
import { deg2rad, rad2deg, StateEvents } from "@cardsgame/utils"
import { Entity } from "./entity"
import { EntityMap } from "./entityMap"
import { Player } from "./player"

export class State extends EventEmitter {
  // 60 cm
  tableWidth = 600
  // 60 cm
  tableHeight = 600

  entities: Entity

  players = new EntityMap<PlayerData>()
  clients = new EntityMap<string>()
  currentPlayerIdx: number
  get currentPlayer(): PlayerData {
    return this.players[this.currentPlayerIdx]
  }
  get playersCount(): number {
    return this.players.length
  }

  // TODO: think this through:
  gameVariants: any

  isGameStarted = false
  @nosync
  _lastID = -1
  @nosync
  _allEntities = new Map<number, Entity>()

  constructor(options?: IStateOptions) {
    super()
    // TODO: do something with these options.

    this.entities = new Entity({
      state: this,
      type: "root",
      name: "root"
    })

    this.setupListeners()
  }

  setupListeners() {
    this.on(StateEvents.privatePropsSyncRequest, (client: string) => {
      // Bubble it down to every entity
      logs.info("State.privatePropsSyncRequest")
      this.entities.emit(StateEvents.privatePropsSyncRequest, client)
    })
  }

  /**
   * Registers new entity to the gamestate
   * @param entity
   * @returns new ID to be assigned to that entity
   */
  rememberEntity(entity: Entity) {
    const newID = ++this._lastID
    this._allEntities.set(newID, entity)
    return newID
  }

  /**
   * Get an Entity by its ID
   * @param id
   */
  getEntity(id: EntityID): Entity
  /**
   * Get an Entity by its idx path
   * @param path
   */
  getEntity(path: number[]): Entity
  getEntity(idOrPath: EntityID | number[]): Entity {
    if (Array.isArray(idOrPath)) {
      const travel = (entity: Entity, path: number[]) => {
        const idx = path.shift()
        const newChild = entity.children[idx] as Entity
        if (!newChild) {
          logs.error("getEntity/path", `This entity doesn't have such child`)
        }
        if (path.length > 0) {
          return travel(newChild, path)
        } else {
          return newChild
        }
      }
      return travel(this.entities, [...idOrPath])
    }
    return this._allEntities.get(idOrPath)
  }

  /**
   * Gets an array of all entities from the top-most parent
   * to the lowest of the child.
   */
  getEntitiesAlongPath(path: number[]): Entity[] {
    const travel = (entity: Entity, path: number[], result: Entity[] = []) => {
      const idx = path.shift()
      const newChild = entity.children[idx] as Entity
      if (!newChild) {
        logs.error("getEntity/path", `This entity doesn't have such child`)
      }
      result.push(newChild)
      if (path.length > 0) {
        return travel(newChild, path, result)
      } else {
        return result
      }
    }
    return travel(this.entities, [...path])
  }

  logTreeState() {
    logs.log("")
    const indent = (level: number) => {
      return "│ ".repeat(level)
    }
    const travel = (entity: Entity, level: number = 0) => {
      entity.childrenArray.map((child, idx, entities) => {
        if (
          child.parentEntity.isContainer &&
          entities.length > 5 &&
          idx < entities.length - 5
        ) {
          // That's too much, man!
          if (idx === 0) {
            logs.log(`${indent(level)}`, "...")
          }
          return
        }

        const owner = child.owner

        const lastChild = entities.length - 1 === idx
        const sIdx = idx === child.idx ? `${idx}` : `e${child.idx}:s${idx}`
        const sVisibility = child.visibleToPublic ? "Pub" : ""

        const sChildren = child.length > 0 ? child.length + " children" : ""
        const sOwner = owner ? `(${owner.type} ${owner.clientID})` : ""
        const branchSymbol = lastChild ? "┕━" : "┝━"

        logs.log(
          `${indent(level)}${branchSymbol}[${sIdx}]${sVisibility}`,
          `${child.type}:${child.name}-[${child.idx}]`,
          sChildren,
          sOwner
        )
        if (child.length > 0) {
          travel(child, level + 1)
        }
      })
    }
    const root = this.entities
    logs.log(
      `┍━[${root.idx}]`,
      `${root.type}:${root.name}`,
      root.length + "children"
    )
    travel(root, 1)
  }

  static ROOT_ID = 0
}

// Get rid of EventEmitter stuff from the client
nosync(State.prototype, "_events")
nosync(State.prototype, "_eventsCount")
nosync(State.prototype, "_maxListeners")
nosync(State.prototype, "domain")

export interface IStateOptions {
  minClients: number
  maxClients: number
  hostID: string
}

export type PlayerData = {
  entity: Entity
  clientID: string
}
