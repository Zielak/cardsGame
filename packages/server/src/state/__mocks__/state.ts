import type { Player } from "../../player/player.js"

export class State<
  V extends Record<string, unknown> = Record<string, unknown>
> {
  name = "Unnamed"
  type = "state"
  tableWidth = 60 // 60 cm
  tableHeight = 60 // 60 cm
  clients = new Array<string>()
  turnBased = true
  round = 0
  players = new Array<Player>()
  currentPlayerIdx = 0

  get currentPlayer(): Player {
    return this.turnBased ? this.players[this.currentPlayerIdx] : null
  }
  isGameStarted = false
  isGameOver = false

  ui = {}

  playerViewPosition = {
    alignX: "center",
    alignY: "bottom",
    paddingX: 0,
    paddingY: 0,
  }

  variantData: V
  hijacksInteractionTarget = false

  id = 0

  _registerEntity(): void {
    // do nothing
  }
}
