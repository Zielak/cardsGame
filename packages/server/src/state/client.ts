import { def } from "@cardsgame/utils"
import { Schema } from "@colyseus/schema"

import { type } from "@/annotations/type.js"

type GameClientOptions = {
  id: string
  ready?: boolean
  isBot?: boolean
}

export class GameClient extends Schema {
  @type("string") id: string
  @type("boolean") ready: boolean
  @type("boolean") isBot: boolean

  constructor(options: GameClientOptions) {
    super()
    this.id = options.id
    this.ready = def(options.ready, false)
    this.isBot = def(options.isBot, false)
  }
}
