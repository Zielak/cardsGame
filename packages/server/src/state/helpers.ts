import { limit, logs } from "@cardsgame/utils"

import { Bot, isBot } from "../players/bot"
import { Player } from "../players/player"
import { IdentityTrait } from "../traits/identity"
import { hasChildren, isParent, ParentTrait } from "../traits/parent"

import { State } from "."

/**
 * Will get you an index of given player in turn queue.
 * Useful if you happen to have just a `Player` reference at hand.
 */
export function getPlayersIndex(state: State, player: Player): number {
  return state.players.findIndex((p) => p === player)
}

export function getNextPlayerIdx(state: State): number {
  const current = limit(state.currentPlayerIdx, 0, state.players.length - 1)
  return current + 1 === state.players.length ? 0 : current + 1
}

export function getNextPlayer(state: State): Player {
  return state.players[getNextPlayerIdx(state)]
}

export function getPreviousPlayerIdx(state: State): number {
  const current = limit(state.currentPlayerIdx, 0, state.players.length - 1)
  return current - 1 === -1 ? state.players.length - 1 : current - 1
}

export function getPreviousPlayer(state: State): Player {
  return state.players[getPreviousPlayerIdx(state)]
}

export function getPlayerByName(state: State, name: string): Player {
  return state.players.find((player) => player.name === name)
}

export function getAllBots(state: State): Bot[] {
  return state.players.filter((player) => isBot(player)) as Bot[]
}

/**
 * Gets an array of all entities from the top-most parent
 * to the lowest of the child.
 */
export function getEntitiesAlongPath(
  state: State,
  path: number[]
): IdentityTrait[] {
  const travel = (
    parent: ParentTrait,
    remainingPath: number[],
    result: any[] = []
  ): any[] => {
    const idx = remainingPath.shift()
    const newChild = parent.getChild(idx)
    if (!newChild) {
      logs.error("getEntitiesAlongPath", `This entity doesn't have such child.`)
      return result
    }

    if (remainingPath.length > 0 && !hasChildren(newChild)) {
      logs.error(
        "getEntitiesAlongPath",
        `Path inaccessible, entity doesn't have any children. Stopped at [${path}]. Remaining path: [${remainingPath}]`
      )
      return result
    }

    result.push(newChild)
    if (remainingPath.length > 0 && isParent(newChild)) {
      return travel(newChild, remainingPath, result)
    } else {
      return result
    }
  }
  return travel(state, [...path])
}
