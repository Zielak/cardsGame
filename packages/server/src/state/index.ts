export * from "./helpers.js"
export * from "./state.js"
