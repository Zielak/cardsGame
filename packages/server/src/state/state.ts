import { MapSchema } from "@colyseus/schema"
import { mapCount } from "@cardsgame/utils"

import { type, containsChildren } from "../annotations"
import { Player } from "../player"
import { PlayerViewPosition } from "../playerViewPosition"

// Trait
import { Entity, applyTraitsMixins } from "../traits/entity"
import { IdentityTrait } from "../traits/identity"
import { LabelTrait } from "../traits/label"
import { ParentArrayTrait } from "../traits/parentArray"

@containsChildren()
@applyTraitsMixins([IdentityTrait, LabelTrait, ParentArrayTrait])
export class State extends Entity<{}> {
  type = "state"

  @type("number") tableWidth = 60 // 60 cm

  @type("number") tableHeight = 60 // 60 cm

  /**
   * IDs of clients who are connecting to the room.
   * A "client" is someone who just stopped by in this room
   * and not necessarily someone who is playing the game.
   */
  @type({ map: "string" }) clients = new MapSchema<string>()

  /**
   * How much clients are connected to the room
   */
  get clientsCount(): number {
    return mapCount(this.clients)
  }

  /**
   * Games are turn-based by default. Each player takes their turns one by one.
   * Set this to `false` to allow simultaneous play.
   * Don't rely on `currentPlayer` value for non turn-based games.
   */
  @type("boolean") turnBased = true

  /**
   * Current round number. Increased using `NextRound` command.
   */
  @type("uint16") round = 0

  /**
   * List of player - game participants, after the game starts.
   */
  @type({ map: Player }) players = new MapSchema<Player>()

  @type("uint8") currentPlayerIdx = 0

  get currentPlayer(): Player {
    return this.turnBased ? this.players[this.currentPlayerIdx] : null
  }

  get playersCount(): number {
    return mapCount(this.players)
  }

  @type("boolean") isGameStarted = false
  @type("boolean") isGameOver = false

  @type({ map: "string" }) ui: StateUI = new MapSchema<string | string[]>()

  /**
   * A construct describing how should player's "focused" items
   * be positioned in his view. Containers of other players
   * will not follow these rules.
   * Default is: center/bottom.
   */
  @type(PlayerViewPosition) playerViewPosition = new PlayerViewPosition()

  /**
   * @private
   */
  _lastID = -1

  /**
   * Use `find()` or `getEntity()` methods instead
   * @private
   */
  _allEntities = new Map<number, IdentityTrait>()

  constructor() {
    super(undefined)

    this.hijacksInteractionTarget = false
  }
}

interface Mixin extends IdentityTrait, LabelTrait, ParentArrayTrait {}

export interface State extends Mixin {}

export interface StateUI {
  clone: () => MapSchema
  onAdd: (item: any, key: string) => void
  onRemove: (item: any, key: string) => void
  onChange: (item: any, key: string) => void
}
