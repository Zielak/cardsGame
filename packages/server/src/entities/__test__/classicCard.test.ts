import { State } from "../../state/state.js"
import { ClassicCard } from "../classicCard.js"

let card: ClassicCard
let state: State

beforeEach(() => {
  state = new State()
})

afterEach(() => {
  card = null
})

describe("constructor", () => {
  test("sets up defaults", () => {
    card = new ClassicCard(state, { suit: "H", rank: "A" })

    expect(card).toMatchSnapshot()
  })
})

describe("state changers", () => {
  beforeEach(() => {
    card = new ClassicCard(state, { suit: "H", rank: "A" })
  })
  test("flip", () => {
    expect(card.faceUp).toBe(false)
    card.flip()
    expect(card.faceUp).toBe(true)
    card.flip()
    expect(card.faceUp).toBe(false)
  })
  test("faceUp", () => {
    expect(card.faceUp).toBe(false)
    card.flipUp()
    expect(card.faceUp).toBe(true)
    card.flipUp()
    expect(card.faceUp).toBe(true)
  })
  test("faceDown", () => {
    expect(card.faceUp).toBe(false)
    card.flipUp()
    expect(card.faceUp).toBe(true)
    card.flipDown()
    expect(card.faceUp).toBe(false)
    card.flipDown()
    expect(card.faceUp).toBe(false)
  })
})
