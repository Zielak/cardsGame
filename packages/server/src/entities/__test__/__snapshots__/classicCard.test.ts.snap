// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`constructor sets up defaults 1`] = `
{
  "angle": 0,
  "faceUp": false,
  "idx": 0,
  "name": "HA",
  "ownersMainFocus": false,
  "rank": "A",
  "suit": "H",
  "type": "classicCard",
  "x": 0,
  "y": 0,
}
`;
